sympy/matrices/matrices.py

```diff
@@ -73,2747 +76,3386 @@ def __repr__(self):
-        the safest option:
+        return all(self[i, j] == dirac(i, j) for i in range(self.rows) for j in
+                   range(self.cols))
 
-        >>> A = Matrix([[1, 2], [x, 0]])
-        >>> A.charpoly().as_expr()
-        _lambda**2 - _lambda - 2*x
-        >>> A.charpoly(x).as_expr()
-        x**2 - 3*x
+    def _eval_is_lower_hessenberg(self):
+        return all(self[i, j].is_zero
+                   for i in range(self.rows)
+                   for j in range(i + 2, self.cols))
 
-        See Also
-        ========
+    def _eval_is_lower(self):
+        return all(self[i, j].is_zero
+                   for i in range(self.rows)
+                   for j in range(i + 1, self.cols))
 
-        berkowitz
-        """
-        return PurePoly(list(map(simplify, self.berkowitz()[-1])), x)
+    def _eval_is_symbolic(self):
+        return self.has(Symbol)
 
-    def berkowitz_det(self):
-        """Computes determinant using Berkowitz method.
+    def _eval_is_symmetric(self, simpfunc):
+        mat = self._new(self.rows, self.cols, lambda i, j: simpfunc(self[i, j] - self[j, i]))
+        return mat.is_zero
 
-        See Also
+    def _eval_is_zero(self):
+        if any(i.is_zero == False for i in self):
+            return False
+        if any(i.is_zero == None for i in self):
+            return None
+        return True
+
+    def _eval_is_upper_hessenberg(self):
+        return all(self[i, j].is_zero
+                   for i in range(2, self.rows)
+                   for j in range(i - 1))
+
+    def _eval_values(self):
+        return [i for i in self if not i.is_zero]
+
+    def atoms(self, *types):
+        """Returns the atoms that form the current object.
+
+        Examples
         ========
 
-        det
-        berkowitz
+        >>> from sympy.abc import x, y
+        >>> from sympy.matrices import Matrix
+        >>> Matrix([[x]])
+        Matrix([[x]])
+        >>> _.atoms()
+        set([x])
         """
-        if not self.is_square:
-            raise NonSquareMatrixError()
-        if not self:
-            return S.One
-        poly = self.berkowitz()[-1]
-        sign = (-1)**(len(poly) - 1)
-        return sign*poly[-1]
 
-    def berkowitz_eigenvals(self, **flags):
-        """Computes eigenvalues of a Matrix using Berkowitz method.
+        types = tuple(t if isinstance(t, type) else type(t) for t in types)
+        if not types:
+            types = (Atom,)
+        return self._eval_atoms(*types)
 
-        See Also
+    @property
+    def free_symbols(self):
+        """Returns the free symbols within the matrix.
+
+        Examples
         ========
 
-        berkowitz
+        >>> from sympy.abc import x
+        >>> from sympy.matrices import Matrix
+        >>> Matrix([[x], [1]]).free_symbols
+        set([x])
         """
-        return roots(self.berkowitz_charpoly(Dummy('x')), **flags)
+        return self._eval_free_symbols()
```

Comment:
What is the reason to have this code in a separate `_eval_*` method? Seems like creating two methods when only one is needed.